### PR TITLE
[WIP] Add Shader Built-in Array Limit

### PR DESCRIPTION
**Pull request by sfricke-samsung**
VUID-VkPipelineShaderStageCreateInfo-maxClipDistances-00708
VUID-VkPipelineShaderStageCreateInfo-maxCullDistances-00709
VUID-VkPipelineShaderStageCreateInfo-maxCombinedClipAndCullDistances-00710
VUID-VkPipelineShaderStageCreateInfo-maxSampleMaskWords-00711

### CODE REVIEW

File: layers/shader_validation.cpp [hunk 1]
```diff
@@ -1306,6 +1306,79 @@ bool CoreChecks::ValidatePushConstantUsage(std::vector<VkPushConstantRange> cons
...
+        auto insn = src->get_def(id);
+        decoration_set decroations = src->get_decorations(insn.word(2));
+
+        // Built-ins are obtained from OpVariable
+        if (((decroations.flags & decoration_set::builtin_bit) != 0) && (insn.opcode() == spv::OpVariable)) {
+            auto type_pointer = src->get_def(insn.word(1));
+            assert(type_pointer.opcode() == spv::OpTypePointer);
+
+            auto type = src->get_def(type_pointer.word(3));
+            if (type.opcode() == spv::OpTypeArray) {
```
Comment (jeffbolznv) [inline layers/shader_validation.cpp:1330]: clip/culldistance are structure members in the vertex shader, so I don't think this code will find them at all. And in some shader stages gl_PerVertex is also an array.

File: layers/shader_validation.cpp [hunk 2]
```diff
@@ -1306,6 +1306,79 @@ bool CoreChecks::ValidatePushConstantUsage(std::vector<VkPushConstantRange> cons
...
+                                             length, phys_dev_props.limits.maxSampleMaskWords,
+                                             report_data->FormatHandle(src->vk_shader_module).c_str());
+                        }
+                        break;
+                }
+            }
+        }
+    }
+
+    if ((clipDistanceLength + cullDistanceLength) > phys_dev_props.limits.maxCombinedClipAndCullDistances) {
```
Comment (jeffbolznv) [inline layers/shader_validation.cpp:1371]: I don't think this check is sound. See internal issues 1406 and 1019 for some history on this. As an example, if you take a vertex shader that just writes `gl_ClipDistance[7] = 1;`, glslang will generate a PerVertex structure with gl_ClipDistance[8] and gl_CullDistance[1].

File: tests/vklayertests_pipeline_shader.cpp [hunk 3]
```diff
@@ -8450,3 +8450,183 @@ TEST_F(VkLayerTest, ValidateGetRayTracingCaptureReplayShaderGroupHandlesKHR) {
...
+    // Exceed combined clip and cull distance array size
+    char const *combinedSource =
+        "#version 450\n"
+        "layout(location = 0) out vec4 uFragColor;\n"
+        "void main(){\n"
+        "   float x = gl_ClipDistance[1];\n"
+        "   float y = gl_CullDistance[2];\n"
+        "   uFragColor = vec4(x,y,0,1);\n"
+        "}\n";
+    VkShaderObj fsCombined(m_device, combinedSource, VK_SHADER_STAGE_FRAGMENT_BIT, this);
```
Comment (jeffbolznv) [inline tests/vklayertests_pipeline_shader.cpp:8625]: Given the issues I mentioned above, I think the tests need to be improved (other shader stages, and probe the limits more closely).

File: layers/shader_validation.cpp [hunk 4]
```diff
@@ -1306,6 +1306,79 @@ bool CoreChecks::ValidatePushConstantUsage(std::vector<VkPushConstantRange> cons
...
+                                       VkShaderStageFlagBits stage) const {
+    bool skip = false;
+
+    // Currently all builtin tested are only found in fragment shaders
+    if (stage != VK_SHADER_STAGE_FRAGMENT_BIT) {
+        return skip;
+    }
+
+    uint32_t clipDistanceLength = 0;
+    uint32_t cullDistanceLength = 0;
```
Comment (jeffbolznv) [inline layers/shader_validation.cpp:1319]: Probably need to separately distinguish between input and output storage class variables.

Comment (sfricke-samsung) [reply layers/shader_validation.cpp:1319]: yep, I see that tess/geo can do both input/output

File: layers/shader_validation.cpp (hunk 2 above)
Comment (sfricke-samsung) [reply layers/shader_validation.cpp:1371]: > The sum of the sizes of any pair of arrays declared with the ClipDistance and CullDistance built-in decoration used by a single shader stage in a shader module must be less than or equal to this limit.

if it generates `gl_ClipDistance[8]` and `gl_CullDistance[1]`  is that not still `9 + 2`

**Comment by mark-lunarg** [comment]
@sfricke, @Tony-LunarG, same as the others, closing this for now -- when there's time later, we can reopen this PR or create a new one -- thanks!

**Pull request closed without merging**